river/drift/kswin.py

```diff
@@ -107,31 +122,29 @@ def update(self, value) -> tuple:
 
         """
         self.n += 1
-        currentLength = self.window.shape[0]
-        if currentLength >= self.window_size:
-            self.window = np.delete(self.window, 0)
-            rnd_window = np.random.choice(
-                self.window[: -self.stat_size], self.stat_size
-            )
 
-            (st, self.p_value) = stats.ks_2samp(
-                rnd_window, self.window[-self.stat_size :], mode="exact"
+        self.window.append(value)
+        if len(self.window) >= self.window_size:
+            rnd_window = [
+                self.window[r]
+                for r in self._rng.sample(
+                    range(self.window_size - self.stat_size), self.stat_size
+                )
+            ]
+            most_recent = list(
+                itertools.islice(
+                    self.window, self.window_size - self.stat_size, self.window_size
+                )
             )
 
+            st, self.p_value = stats.ks_2samp(rnd_window, most_recent)
+
             if self.p_value <= self.alpha and st > 0.1:
                 self._in_concept_change = True
-                self.window = self.window[-self.stat_size :]
+                self.window = collections.deque(most_recent, maxlen=self.window_size)
             else:
                 self._in_concept_change = False
         else:  # Not enough samples in sliding window for a valid test
             self._in_concept_change = False
 
-        self.window = np.concatenate([self.window, [value]])
-
         return self._in_concept_change, self._in_warning_zone
```

Comment:
From what I can tell `self._in_warning_zone` is never `True`, correct?

Reply:
For ADWIN no. It only detects drifts

Reply:
We could check if these private variables are used in other detectors. If that is not the case, I vote for explicitly returning `drift, False`.

Reply:
[This](https://github.com/online-ml/river/blob/main/river/drift/ddm.py) detector seems to operate on two levels. There might be others.

I always considered drift detectors taken for granted, so I don't know much about this research field.